File: src/main/java/net/vidageek/games/regex/task/Tasks.java
package net.vidageek.games.regex.task;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import net.vidageek.games.task.IndexedTask;
import net.vidageek.games.task.Task;

public class Tasks {

	private final List<TaskGroup> taskGroups = new ArrayList<TaskGroup>();

	public void add(final TaskGroup group) {
		taskGroups.add(group);
	}

	public IndexedTask at(final int originalIndex) {
		int index = originalIndex;
		int groupCount = 0;
		while ((groupCount < taskGroups.size()) && (index - taskGroups.get(groupCount).size() >= 0)) {
			index -= taskGroups.get(groupCount).size();
			groupCount++;
		}
		TaskGroup group = taskGroups.get(groupCount);
		return new IndexedTask(new GroupedTask(group, group.task(index)), originalIndex);
	}

	public int size() {
		int res = 0;
		for (TaskGroup group : taskGroups) {
			res += group.size();
		}
		return res;
	}

	public List<IndexedTask> all() {
		return unmodifiableIndexedTasks();
	}

	private List<IndexedTask> unmodifiableIndexedTasks() {
		List<IndexedTask> list = new ArrayList<IndexedTask>();
		int i = 0;
		for (TaskGroup group : taskGroups) {
			for (Task task : group) {
				list.add(new IndexedTask(new GroupedTask(group, task), i++));
			}
		}
		return Collections.unmodifiableList(list);
	}

}

File: src/main/scala/vggames/regex/task/Tasks.scala
package vggames.regex.task

import scala.collection.JavaConverters.seqAsJavaListConverter
import scala.collection.mutable.ListBuffer
import vggames.shared.task.IndexedTask
import vggames.shared.task.GroupedTask

class Tasks {
  val taskGroups = ListBuffer[TaskGroup]()

  def add(group : TaskGroup) : Unit = taskGroups += group

  def at(originalIndex : Int) = {
    var index = originalIndex
    var groupCount = 0
    while ((groupCount < taskGroups.size) && (index - taskGroups(groupCount).size >= 0)) {
      index -= taskGroups(groupCount).size
      groupCount += 1
    }
    val group = taskGroups(groupCount)
    new IndexedTask(new GroupedTask(group, group.task(index)), originalIndex)
  }

  def size = taskGroups.foldLeft(0)(_ + _.size)

  def all : java.util.List[IndexedTask] = unmodifiableIndexedTasks.asJava

  private def unmodifiableIndexedTasks = {
    val list = ListBuffer[IndexedTask]()
    var i = -1
    taskGroups.foreach(group =>
      group.foreach(task =>
        list += new IndexedTask(new GroupedTask(group, task), { i += 1; i })))

    list.toList
  }
}